lib/user/store.js

```diff
@@ -0,0 +1,170 @@
+    if (err) {
+      return cb(err);
+    }
+    user.password = hashed;
+    cb(null, user);
+  });
+}
+
+function Store(datasetId, data) {
+  this.data = data;
+  this.topic = {};
+  this.subscription = {};
+  this.datasetId = datasetId;
+}
+
+/**
+ * Helper function to reset underlying collection for testing
+ */
+Store.prototype.setAll = function(arr) {
+  this.data = arr;
+};
+
+Store.prototype.list = ArrayStore.prototype.list;
+
+Store.prototype.read = function(id) {
+  var user = findById(id, this.data);
+  if (!user) {
+    return q.reject(new Error('User not found'));
+  }
+  return q(cloneAndCleanup(user));
+};
+
+Store.prototype.byUsername = function(username) {
+  var user = findByUsername(username, this.data);
+  if (!user) {
+    return q.reject(new Error('User not found'));
+  }
+  return q(cloneAndCleanup(user));
+};
+
+
+Store.prototype.create = function(user) {
+  user = _.clone(user);
+  user.id = shortid.generate();
+  var self = this;
+  return q.nfcall(setUserPassword, user, user.password).then(function(user) {
+    self.data.push(user);
+    return cloneAndCleanup(user);
+  });
+};
+
+Store.prototype.update = function(user) {
+  var originalUser = findById(user.id, this.data);
+  if (!originalUser) {
+    return q.reject(new Error('User not found'));
+  }
+  delete user.id;
+  _.assign(originalUser, user);
+  // update password since this should only be called from the portal
+  var passwordPromise = user.password ? q.nfcall(setUserPassword, originalUser, user.password) : q(originalUser);
+  return passwordPromise.then(cloneAndCleanup);
+};
+
+Store.prototype.updatePassword = function(username, oldPwd, newPwd) {
+  var user = findByUsername(username, this.data);
+  if (!user) {
+    return q.reject(new Error(verifyErrMessage));
+  }
+  return q.nfcall(hash.verify, oldPwd, user.password).then(function(match) {
+    if (!match) {
+      throw new Error();
+    }
+    return q.nfcall(setUserPassword, user, newPwd);
+  }).then(function(user) {
+    return cloneAndCleanup(user);
+  }).catch(function() {
+    return q.reject(new Error(verifyErrMessage));
+  });
+};
+
+function doVerify(password, user) {
+  // user.password here should be the stored hash
+  return q.nfcall(hash.verify, password, user.password).then(function(match) {
+    if (!match) {
+      throw new Error(verifyErrMessage);
+    } else {
+      return true;
+    }
+  });
+}
+
+Store.prototype.verifyPassword = function(username, password) {
+  var user = findByUsername(username, this.data);
+  if (!user) {
+    return q.reject(new Error(verifyErrMessage));
+  }
+  var delay = 500;
+  return q.delay(delay).then(function() {
+    return doVerify(password, user);
+  });
+};
+
+Store.prototype.delete = ArrayStore.prototype.delete;
+
+Store.prototype.listen = function(topicPrefix, mediator) {
+  var self = this;
+  ArrayStore.prototype.listen.apply(this, arguments);
+  // subscribe to extra topics
+  self.topic.usernameLoad = 'wfm:user:username:read';
```

Comment:
@paolobueno Won't this be duplicated https://github.com/feedhenry-raincatcher/raincatcher-mediator/blob/master/lib/array-store.js#L96